alphagov/paas-admin
# Upgrade dependencies

What
----

There is a problem with a new version of TypeScript that hasn't been
occurring before. It's related to Diagnostics being enabled and reporting
on some [practically non-existing issues][1], such as:

> Object is possibly 'undefined'

This could be avoided by adding the `!` symbol to the object which
tricks the `tsc` to trust the developer.

I see that as a scary practice, which could quickly turn into something
like `font-size: 10px !important;`...

Instead, we're telling `jest`, not to run these diagnostics itself, as
personally I trust that linter and `tsc` itself will be able to handle
it and these don't throw these errors.

Saying that, it makes me think there is some other configuration to
jest, that could allow us keep the diagnostics, as well as act the same
way the linter/tsc do.

Disabling it for now, seems like sane approach.

---

We're switching to newer version of node... Be bold.

At the moment, these two places are out of sync. The lts/Carbon will be
running out of life soon.

No reason just to bump to version 10, if version 12 is available.

---

Nock now comes with pre-defined types as part of the dependency,
therefore having `@types/nock` is no longer a requirement.

As it turns out, upgrading nock has highlighted an issue in our
codebase, where we're providing not-escaped URL into query parameters.

That means, the URL could end up with two question marks in it, which
older browsers could be confused about...

Adding the escaping functionality to hopefully manage the URL properly
from now on.

---

The new version of a client comes in with some types by default. There
is no need for us to try and maintain our own version of these types.

Some of the API has changed, but the release notes were fairly
reflective of that change.

The real issue came in with the try method in our codebase, that for
tests is hiding the actual issue, which was that the client now will
validate the token and the IAT and EXP field of the payload need to be
integers instead of floats. Adding some math.round to make sure this is
not an issue for the tests.

How to review
-------------

- Sanity check
- Deploy to your env
- Invite user - some functionality has changed
- Login with SSO - some functionality has changed
